README.md
# CloudTrax Technical Documentation

*This is preliminary documentation. It applies only to networks running under CloudTrax 4.*

This is technical documentation for CloudTrax-related services. These documents fall into two general areas:

### CloudTrax API ###
Full technical detail of a [RESTful API](./api) that provides low-level access to your CloudTrax network and data. You can configure and fully manage your network and nodes and can build a full replacement of the CloudTrax Dashboard if desired.

### Captive Portal ###

CloudTrax' [Captive Portal](./captive_portal) facilities allow you to host and access splash pages both internally within CloudTrax and externally using UAM. These pages can talk to a variety of authentication services, ranging from a built-in vouchering system, to RADIUS servers, to a backend Authentication Server of your own devising.

### Support/Issues ###

Please contact support@cloudtrax.com or submit a request from https://help.cloudtrax.com in case there are issues or questions.

api/history.md
# History endpoints

These endpoints aggregate traffic-based statistical data for a specified network over a given time period. This provides a historical overview of traffic changes over time. These statistics and time-sliced data can be viewed from several different perspectives:

* for all ssid's in the network
* by node (either singly or per all)
* by client (either singly or per all)

functionality | method | endpoint
--- | --- | ---
[retrieve historical traffic statistics for all ssid's](#network) | GET | `/history/network/<network-id>`
[retrieve historical traffic statistics for all clients](#clients) | GET | `/history/network/<network-id>/clients`
[retrieve historical traffic statistics for a single client](#client) | GET | `/history/network/<network-id>/client/<client-id>`
[retrieve historical traffic statistics for all nodes](#nodes) | GET |` /history/network/<network-id>/nodes`
[retrieve historical traffic statistics for a single node](#node) | GET | `/history/network/<network-id>/node/<node-id>`

#### Reporting time span and sampling frequency
Each of the calls allows an optional query-string argument, `period=<timespan>`, to specify the overall time span during which statistics are to be gathered. Allowable values are:  `2hours`, `day`, `week`, and `month`. If the query-string is omitted, `day` is assumed.

The number of samples gathered per reporting period and the per-each-sample time span depend on the reporting period, as follows:

<a name="sampling-table"></a>
##### samples and intervals per reporting period
reporting period | number of samples | sample span
----- | ----- | -----
`2hours` | 24 | 5 minutes
`day` | 288 | 5 minutes
`week` | 288 | 35 minutes
`month` | 288 | 155 minutes


 <a name="network"></a>
### network
`GET /history/network/<network-id>?period=<period>`

This endpoint provides an aggregation of traffic statistics by ssid for all ssid's in the specified network over the given time period.

<a name="query-string-args"></a>
##### query-string arguments
argument | allowable values | required | note
---- | ---- | ---- | -----
`period` | `2hours`, `day`, `week`, `month` | optional | default is `day`

##### example request
`GET /history/network/12554?period=2hours`

##### example output
Output for the above query. For this endpoint, the sample array is partitioned into "rows" of smaller "time slices" (only two out of 24 being shown here). Each row is a snapshot of both upload and download data sampled over the time span of the sample.

The "clients" object contains client counts for each SSID. The "all" field contains the total unique clients across the entire access point.

In this and the other API calls below, "bup" and "bdown" stand for bytes up and bytes down, respectively (total bytes uploaded and downloaded during the time slice), while "rup" and "rdown" stand for rate up and rate down, respectively (ie, upload and download throughput during the same time span).


````json
{
    "clients": {
        "ssid1": 5,
        "ssid2": 0
    },
    "rows": [
        [
            "2013-11-24T13:50:00Z",
            {
                "ssid1": {
                    "traffic": {
                        "udp": {
                            "bup": 278,
                            "bdown": 193,
                            "rup": 278,
                            "rdown": 193
                        },
                        "tcp": {
                            "bup": 278,
                            "bdown": 193,
                            "rup": 278,
                            "rdown": 193
                        }
                    },
                    "users": 1
                },
                "ssid2": {
                    "traffic": {
                        "netbios": {
                            "bup": 1872,
                            "bdown": 0,
                            "rup": 1872,
                            "rdown": 0
                        }
                    },
                    "users": 1
                }
            }
        ],
        [
            "2013-11-24T13:55:00Z",
            {
                "ssid1": {
                    "traffic": {
                        "udp": {
                            "bup": 278,
                            "bdown": 193,
                            "rup": 278,
                            "rdown": 193
                        },
                        "tcp": {
                            "bup": 278,
                            "bdown": 193,
                            "rup": 278,
                            "rdown": 193
                        },
                        "users": 1
                    }
                }
            }
        ]
    ]
}
````

 <a name="clients"></a>
### clients
`GET /history/network/<network-id>/clients?period=<period>`

This endpoint returns a list of all client devices connected to the specified network, with traffic statistics aggregated over the selected time span.

##### query-string arguments
argument | allowable values | required | note
---- | ---- | ---- | -----
`period` | `2hours`, `day`, `week`, `month` | optional | default is `day`

##### example request
`GET https://api.cloudtrax.com/history/network/12345/clients?period=week`

##### example output
````json
{
  "clients": [
    {
      "1c:99:4c:7b:ae:fe": {
        "name": "android-77b812edd3097a81",
        "name_override": "test overwrite",
        "cid": "31792",
        "signal": {
          "antenna1": -75,
          "antenna2": -76
        },
        "mcs": {
          "tx": 3
        },
        "bitrate": {
          "rx": 6,
          "tx": 60
        },
        "channel_width": "40",
        "band": "5.8",
        "last_seen": "2014-02-10T20:20:17Z",
        "last_name": "Garage OM5P",
        "last_node": "ac:86:74:00:0b:f0",
        "os": "android",
        "os_version": "",
        "blocked": 0,
        "link": "unknown",
        "traffic": {
          "ssid1": {
            "bup": 40,

api/code/python27/cloudtrax/__init__.py

import base64
import hashlib
import hmac
import json
import random
import requests
import string
import time

class CloudTrax(object):

    def __init__(self, key, secret):
        self._key = key
        self._secret = secret
        self.api_server = 'https://api.cloudtrax.com'

    def _build_headers(self, endpoint, data=None):
    
        if (data == None):
            body = '';
        else:
            body = data;
    
        # n log_2(52) bits of entropy, at 8 chars this is 45 bits
        nonce = ''.join([random.choice(string.ascii_letters) for _ in xrange(8)])
        now = int(time.time());

        authorization = "key=" + self._key + ",timestamp=" + str(now) + ",nonce=" + nonce;
        signature = hmac.new(key=self._secret, msg=authorization + endpoint + body, digestmod=hashlib.sha256).hexdigest();

        headers = {
            "Authorization": authorization,
            "Signature": signature,
            "Content-Type": "application/json",
            "OpenMesh-API-Version": "1"
        }

        return headers

    def get(self, endpoint):
        headers = self._build_headers(endpoint);
        responce = requests.get(self.api_server + endpoint, headers=headers)
        return responce.json()

    def post(self, endpoint, data):
        body = json.dumps(data);
        headers = self._build_headers(endpoint, body);
        responce = requests.post(self.api_server + endpoint, headers=headers, data=body)
        return responce.json()

    def delete(self, endpoint):
        headers = self._build_headers(endpoint);
        responce = requests.delete(self.api_server + endpoint, headers=headers)
        return responce.json()

if __name__ == "__main__":
    
    key     = '<key-provided-by-open-mesh>'
    secret  = '<secret-provided-by-open-mesh>'
    
    api = CloudTrax(key,secret)

    # -----------------
    # list all networks
    # -----------------
    
    print "\n-- Network List --"
    result = api.get("/network/list")
    if 'errors' in result:
        print json.dumps(result['errors'])
    else:
        for x in result["networks"]:
            print str(x['id']) + " " + str(x['name'])

    # ----------------------------
    # create a new minimal network
    # ----------------------------
      
    #data = {
    #    'name': 'Howards-test-network-#10',
    #    'password': 'my-beloved-dead-dog-rufus',
    #    'timezone': 'America/Los_Angeles',
    #    'country_code': 'US'
    #}
    
    #del_id = None
    
    #print "\n-- Creating Network --"
    #result = api.post("/network", data)
    #if 'errors' in result:
    #    print json.dumps(result['errors'])
    #else:
    #    print "Created Netrork id# " + str(result['id'])
    #    del_id = result['id']
    
    # -----------------
    # list all networks
    # -----------------
    
    #print "\n-- Network List --"
    #result = api.get("/network/list")
    #if 'errors' in result:
    #    print json.dumps(result['errors'])
    #else:
    #    for x in result["networks"]:
    #        print str(x['id']) + " " + str(x['name'])
        
    # -------------------------------
    # delete the network just created
    # -------------------------------
    
    #print "\n-- Deleting Network --"
    #if ('del_id' in locals()):
    #    result = api.delete("/network/" + str(del_id))
    #    if 'errors' in result:
    #        print json.dumps(result['errors'])
    #    else:
    #        if result['code'] == 1009:
    #            print "Deleted network id# " + str(del_id)
    #        else:
    #            print json.dumps(result)

    # -----------------
    # list all networks
    # -----------------
    
    #print "\n-- Network List --"
    #result = api.get("/network/list")
    #if 'errors' in result:
    #    print json.dumps(result['errors'])
    #else:
    #    for x in result["networks"]:
    #        print str(x['id']) + " " + str(x['name'])


captive_portal/authentication/http/code/php/decode_password.php
<?php

/**
 * decode_password - decode encoded password to ascii string
 * @RA: Request Authenticator passed in from the Access Point
 * @encoded: The encoded password
 * @secret: Shared secret between node and server
 *
 * Returns decoded password or FALSE on error
 */
 
function decode_password($RA, $encoded, $secret)
{
	if (strlen($RA) != 32)
		return FALSE;

	$ra = hex2bin($RA);
	
	if ($ra === FALSE)
		return FALSE;

	if ((strlen($encoded) % 32) != 0)
		return FALSE;

	$bincoded = hex2bin($encoded);

	$password = "";
	$last_result = $ra;

	for ($i = 0; $i < strlen($bincoded); $i += 16) {
		$key = hash('md5', $secret . $last_result, TRUE);
		for ($j = 0; $j < 16; $j++)
			$password .= $key[$j] ^ $bincoded[$i + $j];
		$last_result = substr($bincoded, $i, 16);
	}

	$j = 0;
	for ($i = strlen($password); $i > 0; $i--) {
		if ($password[$i - 1] != "\x00")
			break;
		else
			$j++;
	}

	if ($j > 0) {
		$password = substr($password, 0, strlen($password) - $j);
	}
 
	return $password;
}

/*	given the fixed inputs:
		- $RA (request authenticator)
		- $encoded_pw (the encoded password), and
		- $secret (the shared secret)
	a successful decoding will produce and test for 
	the expected output ($password)
*/

function test_decode_password()
{
	# assume following passed in via the login request from the Access Point:
	$RA = "2590CC8A3930DB222781921A8F8B88B1";
	$encoded_pw = "D8A7B0E4A6122A73705C4640E86CD62EA499201D98C5F436103448C39A537B07";
 
	# and the shared secret
	$secret = "verysecretstring";

	echo '$RA = ' . $RA . "<br/>";
	echo '$encoded_pw = ' . $encoded_pw . "<br/>";
	echo 'strlen($encoded) = ' . strlen($encoded_pw) . "<br/>";
	echo '$secret = ' . $secret . "<br/>";

	$password = decode_password($RA, $encoded_pw, $secret);
	echo "decoded password = " . $password;

	# the decoded password for the above params
	# should equal "123456abcdefghijklmnopqrs" 
	
	if (strcmp($password, "123456abcdefghijklmnopqrs") == 0)
		echo "DECODING WAS SUCCESSFUL";
	else
		echo "DECODING FAILED";
}

test_decode_password();

?>


captive_portal/authentication/http/example_code/php/example_server.php
<?php
header("Content-Type: text/plain");

/**
 * secret - Shared secret between server and node
 */
$secret = "verysecretstring";

/**
 * response - Standard response (is modified depending on the result
 */
$response = array(
	'CODE' => 'REJECT',
	'RA' => '0123456789abcdef0123456789abcdef',
	'BLOCKED_MSG' => 'Rejected! This doesnt look like a valid request',
);


/**
 * print_dictionary - Print dictionary as encoded key-value pairs
 * @dict: Dictionary to print
 */
function print_dictionary($dict)
{
	foreach ($dict as $key => $value) {
		echo '"', rawurlencode($key), '" "', rawurlencode($value), "\"\n";
	}
}

/**
 * calculate_new_ra - calculate new request authenticator based on old ra, code
 *  and secret
 * @dict: Dictionary containing old ra and code. new ra is directly stored in it
 * @secret: Shared secret between node and server
 */
function calculate_new_ra(&$dict, $secret)
{
	if (!array_key_exists('CODE', $dict))
		return;

	$code = $dict['CODE'];

	if (!array_key_exists('RA', $dict))
		return;

	if (strlen($dict['RA']) != 32)
		return;

	$ra = hex2bin($dict['RA']);
	if ($ra === FALSE)
		return;
	
	$dict['RA'] = hash('md5', $code . $ra . $secret);
}

/**
 * decode_password - decode encoded password to ascii string
 * @dict: dictionary containing request RA
 * @encoded: The encoded password
 * @secret: Shared secret between node and server
 *
 * Returns decoded password or FALSE on error
 */
function decode_password($dict, $encoded, $secret)
{
	if (!array_key_exists('RA', $dict))
		return FALSE;

	if (strlen($dict['RA']) != 32)
		return FALSE;

	$ra = hex2bin($dict['RA']);
	if ($ra === FALSE)
		return FALSE;

	if ((strlen($encoded) % 32) != 0)
		return FALSE;

	$bincoded = hex2bin($encoded);

	$password = "";
	$last_result = $ra;

	for ($i = 0; $i < strlen($bincoded); $i += 16) {
		$key = hash('md5', $secret . $last_result, TRUE);
		for ($j = 0; $j < 16; $j++)
			$password .= $key[$j] ^ $bincoded[$i + $j];
		$last_result = substr($bincoded, $i, 16);
	}

	$j = 0;
	for ($i = strlen($password); $i > 0; $i--) {
		if ($password[$i - 1] != "\x00")
			break;
		else
			$j++;
	}

	if ($j > 0) {
		$password = substr($password, 0, strlen($password) - $j);
	}
 
	return $password;
}

/* copy request authenticator */
if (array_key_exists('ra', $_GET) && strlen($_GET['ra']) == 32 && ($ra = hex2bin($_GET['ra'])) !== FALSE && strlen($ra) == 16) {
	$response['RA'] = $_GET['ra'];
}

/* decode password when available */
$password = FALSE;
if (array_key_exists('username', $_GET) && array_key_exists('password', $_GET))
	$password = decode_password($response, $_GET['password'], $secret);

/* store mac when available */
$mac = FALSE;
if (array_key_exists('mac', $_GET))
	$mac = $_GET['mac'];

/* decode request */
if (array_key_exists('type', $_GET)) {
	$type = $_GET['type'];

	switch ($type) {
	case 'login':
		if ($password === FALSE)
			break;

		if ($password == 'ThisIsThePassword' && $_GET['username'] == 'testuser') {
			unset($response['BLOCKED_MSG']);
			$response['CODE'] = "ACCEPT";
			$response['SECONDS'] = 3600;
			$response['DOWNLOAD'] = 2000;
			$response['UPLOAD'] = 800;
		} else {
			$response['BLOCKED_MSG'] = "Invalid username or password";
		}
		break;
	case 'status':
		if ($mac === FALSE)
			break;

		if ($mac == '02:ba:de:af:fe:01') {
			unset($response['BLOCKED_MSG']);
			$response['CODE'] = "ACCEPT";
			$response['SECONDS'] = 120;
			$response['DOWNLOAD'] = 3000;
			$response['UPLOAD'] = 400;
		} else {
			$response['BLOCKED_MSG'] = "Unknown Client";
		}
		break;
	case 'acct':
	case 'logout':
		if ($mac === FALSE)
			break;
		unset($response['BLOCKED_MSG']);
		$response['CODE'] = "OK";
		break;
	};
}

/* calculate new request authenticator based on answer and request -> send it out */
calculate_new_ra($response, $secret);
print_dictionary($response);

?>


api/code/php/simple_api_server_test_harness.php
<?php
	
//  --------------------------
//  replace with the real thing
//  --------------------------

$key = '<key-provided-by-open-mesh>';
$secret = '<secret-provided-by-open-mesh>';

# a fellah needs enums, whatcha gonna do?!
class Method
{
    const GET = 0;
    const POST = 1;
    const PUT = 2;
    const DELETE = 3;
    
    public static function nameForEnumValue($value) {
        switch($value) {
            case 0: return "GET";
            case 1: return "POST";
            case 2: return "PUT";
            case 3: return "DELETE";
        }
    }
};

function print_debug_info($method, $endpoint, $headers) {
    print( "\n" );
    print( "Method: " . Method::nameForEnumValue($method) . "\n");
    print( "Endpoint: " . $endpoint . "\n" );
    print_r($headers);
}

function build_headers($auth, $sign) {
    $headers = array();
    $headers[] = "Authorization: " . $auth;
    $headers[] = "Signature: " . $sign;
    $headers[] = "Content-Type: application/json";
    $headers[] = "OpenMesh-API-Version: 1";
    return $headers;
}

function invoke_curl($method, $endpoint, $headers, $json) {

    $api_server = 'https://api.cloudtrax.com';

    try {
        // get a curl handle then go to town on it

        $ch = curl_init($api_server . $endpoint);
        curl_setopt($ch, CURLOPT_HTTPHEADER, $headers);
        curl_setopt($ch, CURLOPT_RETURNTRANSFER, 1);
    
        if ($method == Method::DELETE)
            curl_setopt($ch, CURLOPT_CUSTOMREQUEST, "DELETE");
        elseif ($method == Method::PUT) {
            curl_setopt($ch, CURLOPT_CUSTOMREQUEST, "PUT");
            curl_setopt($ch, CURLOPT_POSTFIELDS, $json);
        }
        else if ($method == Method::POST) {
            curl_setopt($ch, CURLOPT_POST, 1);
            curl_setopt($ch, CURLOPT_POSTFIELDS, $json);
        }
        
        $result = curl_exec($ch);
        if ($result == FALSE) {
            if (curl_errno($ch) == 0)
                echo "@@@@ NOTE @@@@: nil HTTP return: This API call appears to be broken" . "\n";
            else
                throw new Exception(curl_error($ch), curl_errno($ch));    
        }
        else
            echo "RESULT: \n" . $result . "\n";
    } 
    catch(Exception $e) {
        trigger_error( sprintf('Curl failed with error #%d: "%s"',
            $e->getCode(), $e->getMessage()), E_USER_ERROR);
    }
}

function call_api_server($method, $endpoint, $data) {

    global $key, $secret;
    
    $time = time();
    $nonce = rand();

    if ($method == Method::POST)
        assert( '$data != NULL /* @@@@ POST requires $data @@@@ */');
    elseif ($method == Method::GET || $method == Method::DELETE)
        assert( '$data == NULL /* @@@ GET and DELETE take no $data @@@ */');
        
    $path = $endpoint;
    // if present, concatenate encoded json to $endpoint for Signature:
    if ($data !== NULL) {
    	$json = json_encode($data);
        $path .= $json;
    } else {
        $json = NULL;
    }

    $authorization = "key=" . $key . ",timestamp=" . $time . ",nonce=" . $nonce;
    $signature =  hash_hmac('sha256', $authorization . $path . $json, $secret);
    $headers = build_headers($authorization, $signature);
    
    print_debug_info($method, $endpoint, $headers);
    
    invoke_curl($method, $endpoint, $headers, $json);
}


// 	=====================================================
//	EXAMPLE API ENDPOINT CALLS
// 	=====================================================

// -----------------
// list all networks
// -----------------

call_api_server(Method::GET, "/network/list", NULL);

// ----------------------------
// create a new minimal network
// ----------------------------

$data = array(
            'name' => 'Howards-test-network-#10',
            'password' => 'my-beloved-dead-dog-rufus',
            'timezone' => 'America/Los_Angeles',
            'country_code' => 'US'
        );
call_api_server(Method::POST, "/network", $data);

// ------------------
// add a minimal node
// ------------------

$data = array(
            'mac' => 'AC:86:74:00:22:44',
            'name' => 'Test-Node-#3'
        );

//	initially commented out until we've had an opportunity to 
// 	visually inspect the output of the preceding call to pick 
//	out a valid <network-id> to use

// call_api_server(Method::POST, "/node/network/<network-id>", $data);

// 	----------------------------
//	delete the node just created
//	----------------------------

//	similarly commented out until we've had an opportunity, after
//	running the prior API, to determine the <node-id> of the new node
//	so we can delete it

//call_api_server(Method::DELETE, "/node/<node-id>", NULL);

//	------------------------------------------------------------------
//	get a list of nodes to verify that the one you deleted is gone.

//	(if you're really thorough, you'll run list nodes @before@ you
//	delete it as well to insure it was really there in the first place)
// 	------------------------------------------------------------------

//call_api_server(Method::GET, "/node/network/<network-id>/list", NULL);

// 	----------------------------------------------------------------------
//	update the captive portal 'block' message for SSID 3, network <network-id>
//  ----------------------------------------------------------------------
$data = array( 'ssids' => 
            array( '3' => 
                array( 'captive_portal' => 
                    array( 'block_message' => "Stop in the name of love!" )
                )
            )
        );

//call_api_server(Method::PUT, "/network/<network-id>/settings", $data);

?>
